File: mcts.c
#include "mcts.h"
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <time.h>

// Creates a new MCTS node
MCTSNode *create_node(Card *state, int card_count) {
    MCTSNode *node = (MCTSNode *)malloc(sizeof(MCTSNode));
    for (int i = 0; i < card_count; i++) {
        node->state[i] = state[i];
    }
    node->wins = 0;
    node->visits = 0;
    node->parent = NULL;
    node->child_count = 0;
    return node;
}

// Expands a node with possible moves (draw remaining community cards)
void expand_node(MCTSNode *node, Deck *deck) {
    for (int i = 0; i < 10; i++) { // Simulate 10 possible future states
        Card new_state[7];
        memcpy(new_state, node->state, sizeof(node->state));
        new_state[2 + i] = draw_card(deck); // Simulate community cards

        MCTSNode *child = create_node(new_state, 7);
        child->parent = node;
        node->children[node->child_count++] = child;
    }
}

// Calculates UCB1 value for a node
double ucb1(MCTSNode *node, int parent_visits) {
    if (node->visits == 0) return INFINITY;
    return ((double)node->wins / (double)node->visits) + 
           sqrt(2 * log(parent_visits) / node->visits);
}

// Runs MCTS search from the initial state
MCTSNode *mcts_search(Card *initial_state, Deck *deck) {
    MCTSNode *root = create_node(initial_state, 2);
    clock_t start = clock();

    while (((double)(clock() - start) / CLOCKS_PER_SEC) < 10.0) { // 10 second limit
        MCTSNode *current = root;

        // Selection and Expansion
        if (current->child_count == 0) expand_node(current, deck);

        // Randomly simulate one child
        int random_child = rand() % current->child_count;
        MCTSNode *selected = current->children[random_child];

        // Simulate game outcome
        int result = rand() % 2; // Random win/loss for now
        selected->wins += result;
        selected->visits++;
        current->visits++;
    }

    return root;
}

File: mcts.py
import random
import math
import time

class MCTSNode:
    """
    Represents a node in the MCTS tree.
    Each node tracks wins, visits, and potential child states.
    """

    def __init__(self, state, parent=None):
        self.state = state
        self.parent = parent
        self.children = []
        self.wins = 0
        self.visits = 0

    def ucb1(self):
        """
        Calculates the UCB1 value for this node.
        """
        if self.visits == 0:
            return float('inf')
        return (self.wins / self.visits) + math.sqrt(2 * math.log(self.parent.visits) / self.visits)

    def select_child(self):
        """
        Selects the child with the highest UCB1 value.
        """
        return max(self.children, key=lambda child: child.ucb1())

    def expand(self, possible_moves):
        """
        Expands this node by creating children for possible moves.
        """
        for move in possible_moves:
            child_node = MCTSNode(move, parent=self)
            self.children.append(child_node)

    def backpropagate(self, result):
        """
        Backpropagates the simulation result up the tree.
        """
        self.visits += 1
        self.wins += result
        if self.parent:
            self.parent.backpropagate(result)


class MCTS:
    """
    Monte Carlo Tree Search (MCTS) implementation with UCB1.
    """

    def __init__(self, time_limit=10):
        self.time_limit = time_limit

    def search(self, initial_state, evaluator):
        root = MCTSNode(initial_state)
        start_time = time.time()

        while time.time() - start_time < self.time_limit:
            node = root
            if not node.children:
                node.expand([initial_state])
            result = evaluator(node.state)
            node.children[0].backpropagate(result)

        return root
